File: extensions/xcode/ForgeExtension/ForgeClient.swift
import Foundation

// MARK: - Models

struct BuildError: Codable {
    var file: String
    var message: String
    var severity: String
    var line: Int
    var col: Int
}

private struct ConnectRequest: Encodable {
    let ide: String
    let project_path: String
}

private struct ConnectResponse: Decodable {
    let token: String
    let context_id: String
}

private struct AskRequest: Encodable {
    let context_id: String
    let instruction: String
    let selection: String?
    let file_path: String?
    let stream: Bool
}

private struct AskResponse: Decodable {
    let response: String
}

private struct ErrorsResponse: Decodable {
    let errors: [BuildError]
}

private struct HealthResponse: Decodable {
    let status: String
    let version: String
}

// MARK: - ForgeError

enum ForgeError: LocalizedError {
    case notConnected
    case invalidResponse
    case httpError(Int)
    case decodingError(String)
    case networkError(String)

    var errorDescription: String? {
        switch self {
        case .notConnected:
            return "Not connected to Forge daemon. Make sure `forge start` is running."
        case .invalidResponse:
            return "Invalid response from Forge daemon."
        case .httpError(let code):
            return "Forge daemon returned HTTP \(code)."
        case .decodingError(let detail):
            return "Could not parse daemon response: \(detail)"
        case .networkError(let detail):
            return "Network error: \(detail)"
        }
    }
}

// MARK: - ForgeClient

final class ForgeClient {

    static let shared = ForgeClient()

    private let baseURL = "http://localhost:7878"
    private let session: URLSession
    private let defaults = UserDefaults(suiteName: "dev.forgeapp.ForgeXcode") ?? .standard

    private var token: String? {
        get { defaults.string(forKey: "forge_token") }
        set { defaults.set(newValue, forKey: "forge_token") }
    }

    private var contextID: String? {
        get { defaults.string(forKey: "forge_context_id") }
        set { defaults.set(newValue, forKey: "forge_context_id") }
    }

    private init() {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        session = URLSession(configuration: config)
    }

    // MARK: - Public API

    /// Connect to the daemon and store credentials.
    func connect(projectPath: String, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let url = URL(string: "\(baseURL)/ide/v1/connect") else {
            completion(.failure(ForgeError.invalidResponse))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ConnectRequest(ide: "xcode", project_path: projectPath)
        do {
            request.httpBody = try JSONEncoder().encode(body)
        } catch {
            completion(.failure(error))
            return
        }

        session.dataTask(with: request) { [weak self] data, response, error in
            guard let self else { return }

            if let error {
                completion(.failure(ForgeError.networkError(error.localizedDescription)))
                return
            }

            guard let http = response as? HTTPURLResponse else {
                completion(.failure(ForgeError.invalidResponse))
                return
            }

            guard (200...299).contains(http.statusCode) else {
                completion(.failure(ForgeError.httpError(http.statusCode)))
                return
            }

            guard let data else {
                completion(.failure(ForgeError.invalidResponse))
                return
            }

            do {
                let parsed = try JSONDecoder().decode(ConnectResponse.self, from: data)
                self.token = parsed.token
                self.contextID = parsed.context_id
                completion(.success(()))
            } catch {
                completion(.failure(ForgeError.decodingError(error.localizedDescription)))
            }
        }.resume()
    }

    /// Ask Claude a question with optional code selection and file context.
    func ask(
        instruction: String,
        selection: String?,
        filePath: String?,
        completion: @escaping (Result<String, Error>) -> Void
    ) {
        guard let token, let contextID else {
            completion(.failure(ForgeError.notConnected))
            return
        }

        guard let url = URL(string: "\(baseURL)/ide/v1/ask") else {
            completion(.failure(ForgeError.invalidResponse))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(token, forHTTPHeaderField: "X-Forge-Token")

        let body = AskRequest(
            context_id: contextID,
            instruction: instruction,
            selection: selection,
            file_path: filePath,
            stream: false
        )

        do {
            request.httpBody = try JSONEncoder().encode(body)
        } catch {
            completion(.failure(error))
            return
        }

        session.dataTask(with: request) { data, response, error in
            if let error {
                completion(.failure(ForgeError.networkError(error.localizedDescription)))
                return
            }

            guard let http = response as? HTTPURLResponse else {
                completion(.failure(ForgeError.invalidResponse))
                return
            }

            guard (200...299).contains(http.statusCode) else {
                completion(.failure(ForgeError.httpError(http.statusCode)))
                return
            }

            guard let data else {
                completion(.failure(ForgeError.invalidResponse))
                return
            }

            do {
                let parsed = try JSONDecoder().decode(AskResponse.self, from: data)
                completion(.success(parsed.response))
            } catch {
                // Maybe the daemon returned plain text
                if let text = String(data: data, encoding: .utf8), !text.isEmpty {
                    completion(.success(text))
                } else {
                    completion(.failure(ForgeError.decodingError(error.localizedDescription)))
                }
            }
        }.resume()
    }

    /// Fetch current build errors from the daemon.
    func getErrors(completion: @escaping (Result<[BuildError], Error>) -> Void) {
        guard let token else {
            completion(.failure(ForgeError.notConnected))
            return
        }

        guard let url = URL(string: "\(baseURL)/ide/v1/errors") else {
            completion(.failure(ForgeError.invalidResponse))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue(token, forHTTPHeaderField: "X-Forge-Token")

        session.dataTask(with: request) { data, response, error in
            if let error {
                completion(.failure(ForgeError.networkError(error.localizedDescription)))
                return
            }

            guard let http = response as? HTTPURLResponse else {
                completion(.failure(ForgeError.invalidResponse))
                return
            }

            guard (200...299).contains(http.statusCode) else {
                completion(.failure(ForgeError.httpError(http.statusCode)))
                return
            }

            guard let data else {
                completion(.failure(ForgeError.invalidResponse))
                return
            }

            do {
                let parsed = try JSONDecoder().decode(ErrorsResponse.self, from: data)
                completion(.success(parsed.errors))
            } catch {
                completion(.failure(ForgeError.decodingError(error.localizedDescription)))
            }
        }.resume()
    }

    /// Ping the daemon health endpoint.
    func checkHealth(completion: @escaping (Bool) -> Void) {
        guard let url = URL(string: "\(baseURL)/health") else {
            completion(false)
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 5

        session.dataTask(with: request) { data, response, _ in
            guard
                let http = response as? HTTPURLResponse,
                (200...299).contains(http.statusCode),
                let data,
                let parsed = try? JSONDecoder().decode(HealthResponse.self, from: data),
                parsed.status == "ok"
            else {
                completion(false)
                return
            }
            completion(true)
        }.resume()
    }

    /// True when we have a stored token and context ID.
    var isConnected: Bool { token != nil && contextID != nil }

    /// Ensure connected, connecting first if needed.
    func ensureConnected(projectPath: String, completion: @escaping (Result<Void, Error>) -> Void) {
        if isConnected {
            completion(.success(()))
        } else {
            connect(projectPath: projectPath, completion: completion)
        }
    }
}

File: extensions/android-studio/src/main/kotlin/dev/forgeapp/forge/api/ForgeClient.kt
package dev.forgeapp.forge.api

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

data class ConnectRequest(val ide: String, val project_path: String)
data class ConnectResponse(val token: String, val context_id: String)
data class AskRequest(
    val context_id: String,
    val instruction: String,
    val selection: String? = null,
    val file_path: String? = null,
    val stream: Boolean = false
)
data class AskResponse(val response: String)
data class BuildError(
    val file: String,
    val message: String,
    val line: Int,
    val col: Int,
    val severity: String
)
data class ErrorsResponse(val errors: List<BuildError>)
data class HealthResponse(val status: String, val version: String, val ai_ready: Boolean = false)

class ForgeClient {

    companion object {
        val instance = ForgeClient()
    }

    private val gson = Gson()
    private val JSON = "application/json; charset=utf-8".toMediaType()

    private val client = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val baseUrl = "http://localhost:7878"

    @Volatile private var token: String? = null
    @Volatile private var contextId: String? = null

    // -------------------------------------------------------------------------
    // Public API
    // -------------------------------------------------------------------------

    /**
     * POST /ide/v1/connect
     * Stores the returned token and context_id for subsequent calls.
     */
    fun connect(projectPath: String): Result<Unit> = runCatching {
        val body = gson.toJson(ConnectRequest(ide = "android-studio", project_path = projectPath))
            .toRequestBody(JSON)

        val request = Request.Builder()
            .url("$baseUrl/ide/v1/connect")
            .post(body)
            .build()

        val responseText = executeRequest(request)
        val parsed = gson.fromJson(responseText, ConnectResponse::class.java)
            ?: throw IOException("Empty connect response from daemon")

        token = parsed.token
        contextId = parsed.context_id
    }

    /**
     * POST /ide/v1/ask
     * Requires a prior successful [connect] call.
     */
    fun ask(
        instruction: String,
        selection: String? = null,
        filePath: String? = null
    ): Result<String> = runCatching {
        val ctx = contextId ?: throw IOException("Not connected to daemon. Call connect() first.")
        val tok = token   ?: throw IOException("Not connected to daemon. Call connect() first.")

        val payload = AskRequest(
            context_id  = ctx,
            instruction = instruction,
            selection   = selection,
            file_path   = filePath,
            stream      = false
        )
        val body = gson.toJson(payload).toRequestBody(JSON)

        val request = Request.Builder()
            .url("$baseUrl/ide/v1/ask")
            .addHeader("X-Forge-Token", tok)
            .post(body)
            .build()

        val responseText = executeRequest(request)
        val parsed = gson.fromJson(responseText, AskResponse::class.java)
            ?: throw IOException("Empty ask response from daemon")

        parsed.response
    }

    /**
     * GET /ide/v1/errors
     * Returns current build / lint errors from the daemon.
     */
    fun getErrors(): Result<List<BuildError>> = runCatching {
        val tok = token ?: throw IOException("Not connected to daemon. Call connect() first.")

        val request = Request.Builder()
            .url("$baseUrl/ide/v1/errors")
            .addHeader("X-Forge-Token", tok)
            .get()
            .build()

        val responseText = executeRequest(request)
        val parsed = gson.fromJson(responseText, ErrorsResponse::class.java)
            ?: throw IOException("Empty errors response from daemon")

        parsed.errors
    }

    /**
     * GET /health
     * Does not require authentication.
     */
    fun checkHealth(): Result<HealthResponse> = runCatching {
        val request = Request.Builder()
            .url("$baseUrl/health")
            .get()
            .build()

        val responseText = executeRequest(request)
        gson.fromJson(responseText, HealthResponse::class.java)
            ?: throw IOException("Empty health response from daemon")
    }

    fun isConnected(): Boolean = token != null

    fun disconnect() {
        token     = null
        contextId = null
    }

    // -------------------------------------------------------------------------
    // Internal helpers
    // -------------------------------------------------------------------------

    /**
     * Executes [request] and returns the response body as a String.
     * Throws [IOException] on connection failure or non-2xx HTTP status.
     */
    private fun executeRequest(request: Request): String {
        val response = try {
            client.newCall(request).execute()
        } catch (e: IOException) {
            throw IOException(
                "Cannot reach Forge daemon at $baseUrl. " +
                "Make sure it is running with `forge start`.",
                e
            )
        }

        return response.use { resp ->
            val bodyText = resp.body?.string() ?: ""
            if (!resp.isSuccessful) {
                val detail = try {
                    val errObj = gson.fromJson(bodyText, Map::class.java)
                    errObj?.get("error")?.toString() ?: bodyText
                } catch (_: JsonSyntaxException) {
                    bodyText
                }
                throw IOException("Daemon returned HTTP ${resp.code}: $detail")
            }
            bodyText
        }
    }
}
